pkg/buildid/file.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package buildid
+
+import (
+	"bufio"
+	"bytes"
+	"debug/elf"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"strconv"
+)
+
+var elfPrefix = []byte("\x7fELF")
+
+var readSize = 32 * 1024 // changed for testing
+
+// FromFile reads the build ID from a binary by opening the file,
```

Comment:
This optimization is essentially trying our luck and hoping the compiler laid out the binary that this works? I would like to see quite a few binaries that proves this works. Otherwise I fear we’re just going to end up with wildly different behaviors for pretty unexpected reasons.

Reply:
This is about the header and notes section of the binaries. So according to ELF, this should work. The source of this approach is Go std, and they even use it for all platforms across the board (mach-o, xcoff, etc.). 32kb is a huge chunk of the files to include much information. 

But as we discussed above, this could be an optimization for some edge cases. I'm happy to remove it. If you think it complicates things :)